File: leetcode/124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = -10000000;
    int maxPathSum(TreeNode* root) {
        recursiveHelper(root); 
        return ans;
    }
    
    int recursiveHelper(TreeNode* root) {
        if (root == nullptr) return 0;
        int l = max(recursiveHelper(root->left), 0);
        int r = max(recursiveHelper(root->right), 0);
        ans = max(ans, root->val+l+r);
        return root->val + max(l, r);
    }
};

File: leetcode/124.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var ans int = -1000000;
func maxPathSum(root *TreeNode) int {
    ans = -100000000;
    recursiveHelper(root);
    return ans;
}

func recursiveHelper(root *TreeNode) int {
    if root == nil {
        return 0;
    }
    
    l := max(recursiveHelper(root.Left), 0);
    r := max(recursiveHelper(root.Right), 0);
    ans = max(ans, root.Val+l+r);
    return root.Val + max(l, r);
}

func max(a, b int) int {
    if a > b {
        return a;
    } 
    return b;
}
